lib/byron/src/Cardano/Wallet/Byron/Network.hs

```diff
@@ -424,54 +426,90 @@ mkNetworkClient tr bp chainSyncQ localTxSubmissionQ =
     let tracers = NetworkConnectTracers
             { nctMuxTracer = nullTracer
             , nctHandshakeTracer = contramap MsgHandshakeTracer tr
             }
     let socket = localSnocket iocp addr
-    recovering policy
-        [ const $ Handler handleIOException
-        , const $ Handler handleMuxError
-        ] $ \status -> do
-            traceWith tr $ MsgCouldntConnect (rsIterNumber status)
-            connectTo socket tracers versions addr
+    recovering policy handlers $ \status -> do
+        traceWith tr $ MsgCouldntConnect (rsIterNumber status)
+        connectTo socket tracers versions addr
   where
     -- .25s -> .25s -> .5s → .75s → 1.25s → 2s
     policy :: RetryPolicyM IO
     policy = fibonacciBackoff 250_000 & capDelay 2_000_000
 
-    -- There's a race-condition when starting the wallet and the node at the
-    -- same time: the socket might not be there yet when we try to open it.
-    -- In such case, we simply retry a bit later and hope it's there.
-    handleIOException :: IOException -> IO Bool
-    handleIOException e
-        | isDoesNotExistError e     = pure True
-        | isResourceVanishedError e = do
-            traceWith tr $ MsgConnectionLost (Just e)
-            pure True
-        | otherwise = pure False
-      where
-        isResourceVanishedError = isInfixOf "resource vanished" . show
-
-    -- Recover frmo error when the connection with the node is lost.
-    handleMuxError :: MuxError -> IO Bool
-    handleMuxError = pure . errorType >=> \case
-        MuxUnknownMiniProtocol -> pure False
-        MuxDecodeError -> pure False
-        MuxIngressQueueOverRun -> pure False
-        MuxInitiatorOnly -> pure False
-        MuxIOException e -> handleIOException e
-        MuxBearerClosed -> do
-            traceWith tr $ MsgConnectionLost Nothing
-            pure True
+-- | Handlers that are retrying on every connection lost.
+retryOnConnectionLost
+    :: Tracer IO NetworkLayerLog
+    -> [RetryStatus -> Handler IO Bool]
+retryOnConnectionLost tr =
+    [ const $ Handler $ handleIOException tr' True
+    , const $ Handler $ handleMuxError tr' True
+    ]
+  where
+    tr' = contramap MsgConnectionLost tr
+
+-- | Handlers that are failing if the connection is lost
+failOnConnectionLost
+    :: Tracer IO NetworkLayerLog
+    -> [RetryStatus -> Handler IO Bool]
+failOnConnectionLost tr =
+    [ const $ Handler $ handleIOException tr' False
+    , const $ Handler $ handleMuxError tr' False
+    ]
+  where
+    tr' = contramap MsgConnectionLost tr
 
+-- There's a race-condition when starting the wallet and the node at the
+-- same time: the socket might not be there yet when we try to open it.
+-- In such case, we simply retry a bit later and hope it's there.
+--
+-- When the node's connection vanished, we may also want to handle things in a
+-- slightly different way depending on whether we are a waller worker or just
+-- the node's tip thread.
+handleIOException
```

Comment:
Yeah, this is why I've added the two helper functions to be used at call sites "retryOnConnectionLost" and "failOnConnectionLost" to avoid using Booleans there. I didn't bother however also defining newtypes for the intermediate functions. I'll add a comment though.